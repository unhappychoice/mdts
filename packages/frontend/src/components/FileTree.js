"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const ChevronLeft_1 = __importDefault(require("@mui/icons-material/ChevronLeft"));
const ChevronRight_1 = __importDefault(require("@mui/icons-material/ChevronRight"));
const Description_1 = __importDefault(require("@mui/icons-material/Description"));
const ExpandMore_1 = __importDefault(require("@mui/icons-material/ExpandMore"));
const UnfoldLess_1 = __importDefault(require("@mui/icons-material/UnfoldLess"));
const UnfoldMore_1 = __importDefault(require("@mui/icons-material/UnfoldMore"));
const material_1 = require("@mui/material");
const x_tree_view_1 = require("@mui/x-tree-view");
const SimpleTreeView_1 = require("@mui/x-tree-view/SimpleTreeView");
const react_1 = __importDefault(require("react"));
const useFileTree_1 = require("../hooks/apis/useFileTree");
const renderTreeItems = (tree, onFileSelect, parentPath = '') => {
    return tree.map((item) => {
        if (typeof item === 'string') {
            const fileName = item.split('/').pop();
            return ((0, jsx_runtime_1.jsx)(x_tree_view_1.TreeItem, { itemId: item, label: (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "body2", sx: { fontSize: '0.875rem' }, children: fileName }), icon: (0, jsx_runtime_1.jsx)(Description_1.default, { fontSize: "small" }), onClick: () => onFileSelect(item) }, item));
        }
        else {
            const key = Object.keys(item)[0];
            const value = item[key];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            return ((0, jsx_runtime_1.jsx)(x_tree_view_1.TreeItem, { itemId: currentPath, label: (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center' }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { sx: { mr: 1, fontSize: 'small' }, color: "primary" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "body2", sx: { fontSize: '0.875rem' }, children: key })] }), children: Array.isArray(value) && renderTreeItems(value, onFileSelect, currentPath) }, currentPath));
        }
    });
};
const FileTree = ({ onFileSelect, isOpen, onToggle, selectedFilePath }) => {
    const { fileTree, loading, error } = (0, useFileTree_1.useFileTree)();
    const [expanded, setExpanded] = react_1.default.useState([]);
    const handleExpandAll = () => {
        const allItemIds = [];
        const collectIds = (items, parentPath = '') => {
            items.forEach(item => {
                if (typeof item !== 'string') {
                    const key = Object.keys(item)[0];
                    const currentPath = parentPath ? `${parentPath}/${key}` : key;
                    allItemIds.push(currentPath);
                    collectIds(item[key], currentPath);
                }
            });
        };
        if (fileTree) {
            collectIds(fileTree);
        }
        setExpanded(allItemIds);
    };
    const handleCollapseAll = () => {
        setExpanded([]);
    };
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: {
            width: isOpen ? '300px' : '66px',
            bgcolor: 'background.paper',
            p: isOpen ? 2 : 0.5,
            borderRight: '1px solid',
            borderColor: 'divider',
            minHeight: '100%',
            flexShrink: 0,
        }, children: [(0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '34px', marginTop: isOpen ? '0' : '16px', marginBottom: 2 }, children: [isOpen && ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', flex: 1 }, children: [(0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h6", gutterBottom: true, sx: { flex: 1, marginLeft: 1, marginBottom: 0 }, children: "File Tree" }), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: handleExpandAll, size: "small", "aria-label": "expand all", children: (0, jsx_runtime_1.jsx)(UnfoldMore_1.default, {}) }), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: handleCollapseAll, size: "small", "aria-label": "collapse all", children: (0, jsx_runtime_1.jsx)(UnfoldLess_1.default, {}) })] })), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: onToggle, size: "small", sx: { marginBottom: 0, marginLeft: isOpen ? '0' : '12px' }, children: isOpen ? (0, jsx_runtime_1.jsx)(ChevronLeft_1.default, {}) : (0, jsx_runtime_1.jsx)(ChevronRight_1.default, {}) })] }), isOpen && (loading ? ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100vh' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })) : error ? ((0, jsx_runtime_1.jsxs)(material_1.Typography, { color: "error", children: ["Error: ", error] })) : ((0, jsx_runtime_1.jsx)(SimpleTreeView_1.SimpleTreeView, { defaultCollapseIcon: (0, jsx_runtime_1.jsx)(ExpandMore_1.default, {}), defaultExpandIcon: (0, jsx_runtime_1.jsx)(ChevronRight_1.default, {}), expandedItems: expanded, onExpandedItemsChange: (event, itemIds) => setExpanded(itemIds), sx: { flexGrow: 1, maxWidth: 400, overflowY: 'auto' }, children: renderTreeItems(fileTree, onFileSelect) })))] }));
};
exports.default = FileTree;
