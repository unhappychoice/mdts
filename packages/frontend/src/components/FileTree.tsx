import { ArticleOutlined, Clear as ClearIcon, FolderOutlined } from '@mui/icons-material';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import DescriptionIcon from '@mui/icons-material/Description';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import UnfoldLessIcon from '@mui/icons-material/UnfoldLess';
import UnfoldMoreIcon from '@mui/icons-material/UnfoldMore';
import { Box, CircularProgress, IconButton, InputBase, InputAdornment, Typography } from '@mui/material';
import { TreeItem } from '@mui/x-tree-view';
import { SimpleTreeView } from '@mui/x-tree-view/SimpleTreeView';
import React, { useEffect, useMemo } from 'react';
import { useFileTree } from '../hooks/apis/useFileTree';

interface FileTreeItem {
  [key: string]: (FileTreeItem | string)[];
}

interface FileTreeComponentProps {
  onFileSelect: (path: string) => void;
  isOpen: boolean;
  onToggle: () => void;
  selectedFilePath: string | null;
}

const FileTree: React.FC<FileTreeComponentProps> = ({ onFileSelect, isOpen, onToggle }) => {
  const { fileTree, loading, error } = useFileTree();
  const [expanded, setExpanded] = React.useState<string[]>([]);
  const [searchQuery, setSearchQuery] = React.useState<string>('');

  const filteredTree = useMemo(() => filterTree(fileTree, searchQuery), [fileTree, searchQuery]);

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  const handleExpandAll = () => {
    const allItemIds: string[] = [];
    const collectIds = (items: (FileTreeItem | string)[], parentPath: string = '') => {
      items.forEach(item => {
        if (typeof item !== 'string') {
          const key = Object.keys(item)[0];
          const currentPath = parentPath ? `${parentPath}/${key}` : key;
          allItemIds.push(currentPath);
          collectIds(item[key] as FileTreeItem[] | string[], currentPath);
        }
      });
    };
    if (fileTree) {
      collectIds(fileTree);
    }
    setExpanded(allItemIds);
  };

  const handleCollapseAll = () => {
    setExpanded([]);
  };

  useEffect(() => {
    if (searchQuery) {
      const newExpanded: string[] = [];

      const collectExpandedPaths = (items: (FileTreeItem | string)[], parentPath: string = '') => {
        items.forEach(item => {
          if (typeof item !== 'string') { // It's a folder
            const key = Object.keys(item)[0];
            const value = item[key];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            if (!newExpanded.includes(currentPath)) {
              newExpanded.push(currentPath);
            }
            if (Array.isArray(value)) {
              collectExpandedPaths(value, currentPath);
            }
          }
        });
      };

      if (filteredTree) {
        collectExpandedPaths(filteredTree);
      }
      
      setExpanded(newExpanded);
    } else {
      setExpanded([]);
    }
  }, [searchQuery, filteredTree]);

  return (
    <Box sx={{
      width: isOpen ? '300px' : '66px',
      bgcolor: 'background.paper',
      p: isOpen ? 2 : 0.5,
      borderRight: '1px solid',
      borderColor: 'divider',
      minHeight: '100%',
      flexShrink: 0,
    }}>
      <Box sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '34px', marginTop: isOpen ? '0' : '16px', marginBottom: 2 }}>
        {isOpen && (
          <Box sx={{ display: 'flex', alignItems: 'center', flex: 1 }}>
            <Typography variant="h6" gutterBottom sx={{ flex: 1, marginLeft: 1, marginBottom: 0 }}>
              File Tree
            </Typography>
            <IconButton onClick={handleExpandAll} size="small" aria-label="expand all">
              <UnfoldMoreIcon />
            </IconButton>
            <IconButton onClick={handleCollapseAll} size="small" aria-label="collapse all">
              <UnfoldLessIcon />
            </IconButton>
          </Box>
        )}
        <IconButton onClick={onToggle} size="small" sx={{ marginBottom: 0, marginLeft: isOpen ? '0' : '12px' }}>
          {isOpen ? <ChevronLeftIcon /> : <ChevronRightIcon />}
        </IconButton>
      </Box>
      <Box mb={2}>
        <InputBase
          placeholder="Search files..."
          inputProps={{ 'aria-label': 'search files' }}
          sx={{
            flex: 1,
            width: '100%',
            px: 2,
            py: 0,
            border: '1px solid',
            borderColor: 'divider',
            borderRadius: '4px',
            fontSize: '0.875rem',
          }}
          value={searchQuery}
          onChange={handleSearchChange}
          endAdornment={
            searchQuery && (
              <InputAdornment position="end">
                <IconButton
                  onClick={() => setSearchQuery('')}
                  edge="end"
                  size="small"
                  aria-label="clear search"
                >
                  <ClearIcon sx={{ fontSize: '0.875rem', color: 'divider' }}/>
                </IconButton>
              </InputAdornment>
            )
          }
        />
      </Box>
      {isOpen && (
          loading ? (
            <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100vh' }}>
              <CircularProgress />
            </Box>
          ) : error ? (
            <Typography color="error">Error: {error}</Typography>
          ) : (
            <SimpleTreeView
              defaultCollapseIcon={<ExpandMoreIcon />}
              defaultExpandIcon={<ChevronRightIcon />}
              expandedItems={expanded}
              onExpandedItemsChange={(event, itemIds) => setExpanded(itemIds)}
              sx={{ flexGrow: 1, maxWidth: 400, overflowY: 'auto' }}
            >
              {renderTreeItems(filteredTree, onFileSelect)}
            </SimpleTreeView>
          )
      )}
    </Box>
  );
};

export default FileTree;

const renderTreeItems = (
  tree: (FileTreeItem | string)[],
  onFileSelect: (path: string) => void,
  parentPath: string = ''
) => {
  return tree.map((item) => {
    if (typeof item === 'string') {
      const fileName = item.split('/').pop();
      return (
        <TreeItem
          key={item}
          itemId={item}
          label={
            <Box sx={{ display: 'flex', alignItems: 'center' }}>
              <ArticleOutlined sx={{ mr: 1, fontSize: 'small' }} />
              <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
                {fileName}
              </Typography>
            </Box>
          }
          onClick={() => onFileSelect(item)}
        />
      );
    } else {
      const key = Object.keys(item)[0];
      const value = item[key];
      const currentPath = parentPath ? `${parentPath}/${key}` : key;
      return (
        <TreeItem
          key={currentPath}
          itemId={currentPath}
          label={
            <Box sx={{ display: 'flex', alignItems: 'center' }}>
              <FolderOutlined sx={{ mr: 1, fontSize: 'small' }} color="primary" />
              <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
                {key}
              </Typography>
            </Box>
          }
        >
          {Array.isArray(value) && value.length > 0 && renderTreeItems(value, onFileSelect, currentPath)}
        </TreeItem>
      );
    }
  });
};

const filterTree = (tree: (FileTreeItem | string)[], searchQuery: string): (FileTreeItem | string)[] => {
  if (!searchQuery) return tree; // If no search query, return the original tree

  return tree
    .reduce((acc: (FileTreeItem | string)[], item) => {
      if (typeof item === 'string') {
        const fileName = item.split('/').pop() || '';
        return fileName.toLowerCase().includes(searchQuery.toLowerCase())
          ? [...acc, item]
          : acc;
      } else {
        const key = Object.keys(item)[0];
        const value = item[key];

        const children = Array.isArray(value)
          ? filterTree(value, searchQuery)
          : [];

        return children.length > 0
          ? [...acc, { [key]: children } as FileTreeItem ]
          :acc;
      }
    }, []);
};
