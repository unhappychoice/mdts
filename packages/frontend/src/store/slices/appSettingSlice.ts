import { createSlice } from '@reduxjs/toolkit';

type DarkMode = 'dark' | 'light' | 'auto';
type ContentMode = 'full' | 'compact';

interface AppSettingState {
  darkMode: DarkMode;
  contentMode: ContentMode;
  fileTreeOpen: boolean;
  outlineOpen: boolean;
}

const initialState: AppSettingState = {
  darkMode: 'auto',
  contentMode: 'compact',
  fileTreeOpen: true,
  outlineOpen: true,
};

const appSettingSlice = createSlice({
  name: 'appSetting',
  initialState,
  reducers: {
    saveAppSetting: (state, action: PayloadAction<{ darkMode: DarkMode, contentMode: ContentMode }>) => {
      state.darkMode = action.payload.darkMode;
      state.contentMode = action.payload.contentMode;
    },
    toggleFileTree: (state) => {
      state.fileTreeOpen = !state.fileTreeOpen;
    },
    toggleOutline: (state) => {
      state.outlineOpen = !state.outlineOpen;
    },
  },
});

export const { saveAppSetting, toggleFileTree, toggleOutline } = appSettingSlice.actions;

export default appSettingSlice.reducer;
