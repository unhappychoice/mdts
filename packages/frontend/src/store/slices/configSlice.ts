import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export const fetchConfig = createAsyncThunk('config/fetchConfig', async () => {
  return await fetchData<ConfigState>('/api/config', 'json');
});

export const saveConfigToBackend = createAsyncThunk(
  'config/saveConfigToBackend',
  async (config: ConfigState, { dispatch }) => {
    try {
      await fetch('/api/config', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
      });
      dispatch(fetchConfig()); // Re-fetch config after saving
    } catch (error) {
      console.error('Failed to save settings:', error);
      throw error; // Re-throw to allow error handling in component
    }
  }
);

export interface ConfigState {
  fontFamily: string;
  fontFamilyMonospace: string;
  fontSize: number;
  syntaxHighlighterTheme: string;
}

const initialState: ConfigState = {
  fontFamily: 'Roboto',
  fontFamilyMonospace: 'monospace',
  fontSize: 14,
  syntaxHighlighterTheme: 'auto',
};

const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action: PayloadAction<ConfigState>) => {
      state.fontFamily = action.payload.fontFamily;
      state.fontFamilyMonospace = action.payload.fontFamilyMonospace;
      state.fontSize = action.payload.fontSize;
      state.syntaxHighlighterTheme = action.payload.syntaxHighlighterTheme;
    });
  },
});

export default configSlice.reducer;
