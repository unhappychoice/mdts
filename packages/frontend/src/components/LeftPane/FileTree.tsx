import { Box } from '@mui/material';
import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchFileTree, setExpandedNodes, setSearchQuery, expandAllNodes, FileTreeItem } from '../../store/slices/fileTreeSlice';
import { AppDispatch, RootState } from '../../store/store';
import FileTreeHeader from './FileTreeHeader';
import FileTreeSearch from './FileTreeSearch';
import FileTreeContent from './FileTreeContent';

interface FileTreeComponentProps {
  onFileSelect: (path: string) => void;
  isOpen: boolean;
  onToggle: () => void;
  selectedFilePath: string | null;
}

const FileTree: React.FC<FileTreeComponentProps> = ({ onFileSelect, isOpen, onToggle }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { fileTree, filteredFileTree, loading, error, searchQuery, expandedNodes } = useSelector((state: RootState) => state.fileTree);

  useEffect(() => {
    dispatch(fetchFileTree());
  }, [dispatch]);

  const handleSearchChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setSearchQuery(event.target.value));
  }, [dispatch]);

  const handleClearSearch = useCallback(() => {
    dispatch(setSearchQuery(''));
  }, [dispatch]);

  const handleExpandAllClick = useCallback(() => {
    dispatch(expandAllNodes(fileTree));
  }, [dispatch, fileTree]);

  const handleCollapseAll = useCallback(() => {
    dispatch(setExpandedNodes([]));
  }, [dispatch]);

  const handleExpandedItemsChange = useCallback((event: React.SyntheticEvent, itemIds: string[]) => {
    dispatch(setExpandedNodes(itemIds));
  }, [dispatch]);

  useEffect(() => {
    if (!searchQuery) return;

    const newExpanded: string[] = [];

    const collectExpandedPaths = (items: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[], parentPath: string = '') => {
      items.forEach(item => {
        if (!('path' in item)) { // It's a folder
          const key = Object.keys(item)[0];
          const value = (item as { [key: string]: (FileTreeItem | object)[] })[key];
          const currentPath = parentPath ? `${parentPath}/${key}` : key;
          if (!newExpanded.includes(currentPath)) {
            newExpanded.push(currentPath);
          }
          if (Array.isArray(value)) {
            collectExpandedPaths(value, currentPath);
          }
        }
      });
    };

    if (filteredFileTree) {
      collectExpandedPaths(filteredFileTree);
    }

    dispatch(setExpandedNodes(newExpanded));
  }, [searchQuery, filteredFileTree, dispatch]);

  return (
    <Box sx={{
      width: isOpen ? '300px' : '66px',
      bgcolor: 'background.paper',
      py: 2,
      borderRight: '1px solid',
      borderColor: 'divider',
      minHeight: '100%',
      flexShrink: 0,
    }}>
      <FileTreeHeader
        isOpen={isOpen}
        onToggle={onToggle}
        onExpandAllClick={handleExpandAllClick}
        onCollapseAll={handleCollapseAll}
      />
      {isOpen && (
        <FileTreeSearch
          searchQuery={searchQuery}
          onSearchChange={handleSearchChange}
          onClearSearch={handleClearSearch}
        />
      )}
      {isOpen && (
        <FileTreeContent
          filteredFileTree={filteredFileTree}
          loading={loading}
          error={error}
          expandedNodes={expandedNodes}
          onFileSelect={onFileSelect}
          onExpandedItemsChange={handleExpandedItemsChange}
          dispatch={dispatch}
        />
      )}
    </Box>
  );
};

export default FileTree;

