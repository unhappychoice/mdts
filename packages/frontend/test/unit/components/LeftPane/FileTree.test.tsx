import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { thunk } from 'redux-thunk';
import FileTree from '../../../../src/components/LeftPane/FileTree';
import { fetchFileTree, setSearchQuery, expandAllNodes, setExpandedNodes } from '../../../../src/store/slices/fileTreeSlice';

const mockStore = configureStore([thunk]);

jest.mock('../../../../src/store/slices/fileTreeSlice', () => ({
  ...jest.requireActual('../../../../src/store/slices/fileTreeSlice'),
  fetchFileTree: Object.assign(jest.fn(() => ({ type: 'fileTree/fetchFileTree/pending' })), {
    fulfilled: jest.fn((payload) => ({ type: 'fileTree/fetchFileTree/fulfilled', payload })),
  }),
  expandAllNodes: jest.fn((payload) => ({ type: 'fileTree/expandAllNodes', payload })),
  setExpandedNodes: jest.fn((payload) => ({ type: 'fileTree/setExpandedNodes', payload })),
  setFilteredFileTree: jest.fn((payload) => ({ type: 'fileTree/setFilteredFileTree', payload }))
}));

describe('FileTree', () => {
  let store;
  const initialState = {
    fileTree: {
      fileTree: [
        { path: 'file3.js', status: ' ' },
        { 'folder1': [{ path: 'folder1/file1.md', status: ' ' }, { path: 'folder1/file2.txt', status: ' ' }] },
      ],
      filteredFileTree: [
        { path: 'file3.js', status: ' ' },
        { 'folder1': [{ path: 'folder1/file1.md', status: ' ' }, { path: 'folder1/file2.txt', status: ' ' }] },
      ],
      loading: false,
      error: null,
      searchQuery: '',
      expandedNodes: [],
      mountedDirectoryPath: '',
    },
  };

  beforeEach(() => {
    store = mockStore(initialState);
    jest.spyOn(store, 'dispatch');
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('dispatches fetchFileTree on mount', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );
    expect(store.dispatch).toHaveBeenCalledWith(fetchFileTree());
  });

  test('dispatches setSearchQuery on search input change', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const searchInput = screen.getByPlaceholderText('Search files...');
    fireEvent.change(searchInput, { target: { value: 'file' } });

    expect(store.dispatch).toHaveBeenCalledWith(setSearchQuery('file'));
  });

  test('clears search query when clear button is clicked', () => {
    store = mockStore({
      fileTree: {
        ...initialState.fileTree,
        searchQuery: 'file',
        filteredFileTree: [{ path: '/folder1/file1.md', status: ' ' }],
      },
    });
    jest.spyOn(store, 'dispatch');

    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const clearButton = screen.getByLabelText('clear search');
    fireEvent.click(clearButton);

    expect(store.dispatch).toHaveBeenCalledWith(setSearchQuery(''));
  });

  test('dispatches expandAllNodes when expand all button is clicked', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const expandAllButton = screen.getByLabelText('expand all');
    fireEvent.click(expandAllButton);

    expect(store.dispatch).toHaveBeenCalledWith(expandAllNodes(initialState.fileTree.fileTree));
  });

  test('dispatches setExpandedNodes with empty array when collapse all button is clicked', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const collapseAllButton = screen.getByLabelText('collapse all');
    fireEvent.click(collapseAllButton);

    expect(store.dispatch).toHaveBeenCalledWith(setExpandedNodes([]));
  });

  test('updates expandedNodes when search query changes and filteredFileTree is not empty', async () => {
    const updatedFileTree = [{ 'folder1': [{ path: 'folder1/file1.md', status: ' ' }] }];
    let currentStore = mockStore(initialState);
    const { rerender } = render(
      <Provider store={currentStore}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    await act(async () => {
      // Simulate dispatching setSearchQuery and the reducer updating the state
      const newStoreState = {
        fileTree: {
          ...initialState.fileTree,
          searchQuery: 'file1',
          filteredFileTree: updatedFileTree,
        },
      };
      currentStore = mockStore(newStoreState);
      jest.spyOn(currentStore, 'dispatch'); // Re-spy on the new store's dispatch
      rerender(
        <Provider store={currentStore}>
          <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
        </Provider>
      );
    });

    await waitFor(() => {
      expect(currentStore.dispatch).toHaveBeenCalledWith({
        type: 'fileTree/setExpandedNodes',
        payload: ['folder1'],
      });
    });
  });

  test('does not update expandedNodes when search query is empty', async () => {
    const updatedState = {
      fileTree: {
        ...initialState.fileTree,
        searchQuery: '',
        filteredFileTree: [{ 'folder1': [{ path: 'file1.md', status: ' ' }] }],
      },
    };
    store = mockStore(updatedState);
    jest.spyOn(store, 'dispatch');

    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    // Simulate search query change to empty
    store.dispatch(setSearchQuery(''));

    await waitFor(() => {
      expect(store.dispatch).not.toHaveBeenCalledWith({
        type: 'fileTree/setExpandedNodes',
        payload: [],
      });
    });
  });
});
