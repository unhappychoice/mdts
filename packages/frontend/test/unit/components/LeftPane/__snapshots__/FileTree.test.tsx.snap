// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FileTree renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-9681nh"
  >
    <div
      class="MuiBox-root css-xonpca"
    >
      <div
        class="MuiBox-root css-1g3qtqm"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1b6r7r0-MuiTypography-root"
        >
          File Tree
        </h6>
        <button
          aria-label="expand all"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="UnfoldMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
        <button
          aria-label="collapse all"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="UnfoldLessIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 18.59 8.83 20 12 16.83 15.17 20l1.41-1.41L12 14zm9.18-13.18L15.17 4 12 7.17 8.83 4 7.41 5.41 12 10z"
            />
          </svg>
        </button>
      </div>
      <button
        aria-label="toggle file tree"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1nyqon6-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiBox-root css-kuuj3l"
    >
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-pe16f2-MuiInputBase-root"
      >
        <input
          aria-label="search files"
          class="MuiInputBase-input css-e77w8q-MuiInputBase-input"
          placeholder="Search files..."
          type="text"
          value=""
        />
      </div>
    </div>
    <ul
      aria-multiselectable="false"
      class="custom-scrollbar MuiSimpleTreeView-root css-5c3a1o-MuiSimpleTreeView-root"
      defaultcollapseicon="[object Object]"
      defaultexpandicon="[object Object]"
      id="mui-tree-view-1"
      role="tree"
      style="--TreeView-itemChildrenIndentation: 12px;"
    >
      <li
        aria-selected="false"
        class="MuiTreeItem-root MuiSimpleTreeView-root css-jq3mgl-MuiTreeItem-root"
        id="mui-tree-view-1-file3.js"
        role="treeitem"
        style="--TreeView-itemDepth: 0;"
        tabindex="0"
      >
        <div
          class="MuiTreeItem-content MuiSimpleTreeView-itemContent css-cn2zdx-MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer MuiSimpleTreeView-itemIconContainer css-bt3rqe-MuiTreeItem-iconContainer"
          />
          <div
            class="MuiTreeItem-label MuiSimpleTreeView-itemLabel css-1yq7dn0-MuiTreeItem-label"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-eep7r1-MuiSvgIcon-root"
                data-testid="ArticleOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
                />
                <path
                  d="M14 17H7v-2h7zm3-4H7v-2h10zm0-4H7V7h10z"
                />
              </svg>
              <p
                class="MuiTypography-root MuiTypography-body2 css-iewwb3-MuiTypography-root"
              >
                file3.js
              </p>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-expanded="false"
        aria-selected="false"
        class="MuiTreeItem-root MuiSimpleTreeView-root css-jq3mgl-MuiTreeItem-root"
        id="mui-tree-view-1-folder1"
        role="treeitem"
        style="--TreeView-itemDepth: 0;"
        tabindex="-1"
      >
        <div
          class="MuiTreeItem-content MuiSimpleTreeView-itemContent css-cn2zdx-MuiTreeItem-content"
        >
          <div
            class="MuiTreeItem-iconContainer MuiSimpleTreeView-itemIconContainer css-bt3rqe-MuiTreeItem-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="TreeViewExpandIconIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
          <div
            class="MuiTreeItem-label MuiSimpleTreeView-itemLabel css-1yq7dn0-MuiTreeItem-label"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-muidwl-MuiSvgIcon-root"
                data-testid="FolderOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m9.17 6 2 2H20v10H4V6zM10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8z"
                />
              </svg>
              <p
                class="MuiTypography-root MuiTypography-body2 css-1jo9ady-MuiTypography-root"
              >
                folder1
              </p>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
